pkg/apis/batch/validation/validation.go

```diff
@@ -486,9 +499,63 @@ func validatePodTemplateUpdate(spec, oldSpec batch.JobSpec, fldPath *field.Path,
+		// (compared to the oldJob). This allows proceeding with status updates unrelated to the fields violating the
+		// checks, while blocking bad status updates for jobs with correct status.
+		if isConditionTrue(job.Status.Conditions, batch.JobComplete) && isConditionTrue(job.Status.Conditions, batch.JobFailed) {
+			if !isConditionTrue(oldJob.Status.Conditions, batch.JobComplete) || !isConditionTrue(oldJob.Status.Conditions, batch.JobFailed) {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("conditions"), field.OmitValueType{}, "both Complete=True and Failed=True conditions cannot be set"))
+			}
+		}
+		if oldJob.Status.CompletedIndexes != job.Status.CompletedIndexes {
+			if ptr.Deref(job.Spec.CompletionMode, batch.NonIndexedCompletion) == batch.NonIndexedCompletion && job.Status.CompletedIndexes != "" {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("completedIndexes"), job.Status.CompletedIndexes, "cannot set non-empty completedIndexes when non-indexed completion mode"))
+			}
+			if job.Spec.Completions != nil {
+				if err := validateIndexesFormat(job.Status.CompletedIndexes, int(*job.Spec.Completions)); err != nil {
+					allErrs = append(allErrs, field.Invalid(statusFld.Child("completedIndexes"), job.Status.CompletedIndexes, fmt.Sprintf("error parsing completedIndexes: %s", err.Error())))
+				}
+			}
+		}
+		if !ptr.Equal(oldJob.Status.CompletionTime, job.Status.CompletionTime) || isConditionTrue(oldJob.Status.Conditions, batch.JobComplete) != isConditionTrue(job.Status.Conditions, batch.JobComplete) {
+			if job.Status.CompletionTime != nil && !isConditionTrue(job.Status.Conditions, batch.JobComplete) {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("completionTime"), job.Status.CompletionTime, "cannot set completionTime when there is no Complete=True condition"))
+			}
+		}
+		if !ptr.Equal(oldJob.Status.FailedIndexes, job.Status.FailedIndexes) {
+			if job.Spec.BackoffLimitPerIndex == nil && job.Status.FailedIndexes != nil {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("failedIndexes"), *job.Status.FailedIndexes, "cannot set non-null failedIndexes when backoffLimitPerIndex is null"))
+			}
+			if job.Spec.Completions != nil && job.Status.FailedIndexes != nil {
+				if err := validateIndexesFormat(*job.Status.FailedIndexes, int(*job.Spec.Completions)); err != nil {
+					allErrs = append(allErrs, field.Invalid(statusFld.Child("failedIndexes"), job.Status.FailedIndexes, fmt.Sprintf("error parsing failedIndexes: %s", err.Error())))
+				}
+			}
+		}
+		if oldJob.Status.Active != job.Status.Active || !ptr.Equal(oldJob.Status.Ready, job.Status.Ready) {
+			if job.Status.Ready != nil && *job.Status.Ready > job.Status.Active {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("ready"), *job.Status.Ready, "cannot set more ready pods than active"))
+			}
+		}
+	}
```

Comment:
What about comparing `StartTime` vs `CompletionTime`?

Reply:
And `UncountedTerminatedPods`?

Reply:
And `Terminating`?

Reply:
And `Ready`?

Reply:
And `Failed`?

Reply:
The ints can be just ensured they are unset or hold a non-negative value. In general, since we're opening it up, it's better to have a bit tighter validation in place, which if needed we can open up, rather than end up in situation where we'll be fixing bugs coming from 3rd party controllers.

Reply:
> The ints can be just ensured they are unset or hold a non-negative value.

The basic validation of the fields (like being non-negative when set) is already done [here](https://github.com/kubernetes/kubernetes/blob/d2b4928669c633cffb0e4aa6317d0e016ee37de6/pkg/apis/batch/validation/validation.go#L393-L429).

> In general, since we're opening it up, it's better to have a bit tighter validation in place, which if needed we can open up, rather than end up in situation where we'll be fixing bugs coming from 3rd party controllers.

Correct, this is the point of the extra validations. The complication is that we could validate really a lot based on the current Job controller behavior :) For now I only applied validations for things that explicitly follow from the definition of the fields (API comments), or agreed in KEP.

These are some floating ideas for additional validations (not covered yet):
1. `CompletionTime` is set when `Complete=true` 
2. `CompletionTime` is immutable once set
3. `StartTime` is set when we have terminal condition
4. timestamps of `StartTime <= CompletionTime`
5. `Failed` counter is non-decreasing (`Failed` is incremented by the pods removed from `UncountedTerminatedPods.Failed`, ie. `|oldJob.UncountedTerminatedPods.Failed-job.UncountedTerminatedPods.Failed|`)
6. `Succeeded` counter equals `|CompletedIndexes|` for indexed jobs
7. `Failed` counter equals `|FailedIndexes|` for indexed jobs

They all sound reasonable, and are satisfied by the current implementation, but I couldn't find a clear support for them in the API comments. I think adding 1. and 2. might be important as they are directly used by CronJob, but would like to hear your thoughts before making the changes

Reply:
I agree with Maciej in that we should validate as much as possible based on the current behavior of the job-controller.

Reply:
Thanks for the input, makes sense.

I'm still thinking about the time-based check (4.) `StartTime <= CompletionTime`, this might be tricky even for the current built-in controller. Suppose we have a job running job with `StartTime != nil`, then the leader `kube-controller-manager` changes and we have a new one running with a 1min shifted clock. In the meanwhile the job completes, so we could have the back in time `completionTime`. I think such a big clock shift is probably very unlikely, but not sure we make a rule out of this.

Reply:
The controller will keep retrying until `time.Now() >= StartTime`

Reply:
Ok, I just pushed a commit with the strenghtened validation, new cases covered:
1. CompletionTime is set when Complete=true
2. CompletionTime is immutable once set
3. StartTime is set when we have terminal condition
4. timestamps of StartTime <= CompletionTime
5. Failed counter is non-decreasing
6. Succeeded counter is non-decreasing for non-indexed jobs (for indexed it may decrease after scale down)
7. When the job has terminal condition then: active=0, terminating=0, uncountedTerminatingPods is empty

I think 7 is especially important to notify authors of other controllers of leaking pods).

Reply:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123273/pull-kubernetes-e2e-kind-ipv6/1763112742656413696

this failure is related because the built in controller attempted to update the start time for an suspended job. I guess we should just say it cannot be modified to older time not removed. I will update the PR. However this failure demonstrates that some consequences of too strong validation can be hard to predict.

Reply:
> However this failure demonstrates that some consequences of too strong validation can be hard to predict.

The amount of validation logic is quite substantial now and increases a risk of introducing a regression. 

I wonder if we could/should add the validation behind a feature gate? cc @deads2k

And it might be worth considering if scaling this feature back to alpha would have a benefit?

Reply:
> I wonder if we could/should add the validation behind a feature gate? cc @deads2k

This validation is behind  featuregate.  This validation is well-timed to protect the API contract for consumers of our status API so that JobStatus consumers can expect congruent API behavior no matter which manager handles the job.

Reply:
oh - my bad, I have somehow skipped the line with the feature gate check...

Reply:
> And it might be worth considering if scaling this feature back to alpha would have a benefit?

we are going for alpha anyway due to the recent decision to use field instead of label